verilog/analysis/checkers/forbid_implicit_declarations_rule_test.cc

```diff
@@ -0,0 +1,153 @@
+      {"module m;\nwire b;assign {", {kToken, "a"}, ", b,", {kToken, "c"}, "} = 3'b010;\nendmodule"},
+
+      // around scope
+      {"module m;assign ", {kToken, "a1"}, " = 1'b1;\nbegin\nend\nendmodule"},
+      {"module m;begin\nend\nassign ", {kToken, "a1"}, " = 1'b1;endmodule"},
+
+      // declaration and assignement separated by begin-end block
+      {"module m;\nwire a1;\nbegin\nend\nassign a1 = 1'b1;\nendmodule"},
+
+      // out-of-scope
+      {"module m;\nbegin wire a1;\nend\nassign ", {kToken, "a1"}, " = 1'b1;\nendmodule"},
+      {"module m;\nbegin wire a1;\nassign a1 = 1'b0;\nend\nassign ", {kToken, "a1"}, " = 1'b1;\nendmodule"},
+      {"module m;\nwire a1;begin assign ", {kToken, "a1"}, " = 1'b0;\nend\nassign a1 = 1'b1;\nendmodule"},
+
+      // multi-level begin-end blocks
+      {"module m;\n"
+       "  wire x1;\n"
+       "  begin\n"
+       "    wire x2;\n"
+       "    begin\n"
+       "      wire x3;\n"
+       "      begin\n"
+       "        wire x4;\n"
+       "        assign x4 = 1'b0;\n"
+       "        assign ", {kToken, "x3"}, " = 1'b0;\n"
+       "        assign ", {kToken, "x2"}, " = 1'b0;\n"
+       "        assign ", {kToken, "x1"}, " = 1'b0;\n"
+       "      end\n"
+       "      assign ", {kToken, "x4"}, " = 1'b0;\n"
+       "      assign x3 = 1'b1;\n"
+       "      assign ", {kToken, "x2"}, " = 1'b0;\n"
+       "      assign ", {kToken, "x1"}, " = 1'b0;\n"
+       "    end\n"
+       "    assign ", {kToken, "x4"}, " = 1'b0;\n"
+       "    assign ", {kToken, "x3"}, " = 1'b0;\n"
+       "    assign x2 = 1'b0;\n"
+       "    assign ", {kToken, "x1"}, " = 1'b0;\n"
+       "  end\n"
+       "  assign ", {kToken, "x4"}, " = 1'b0;\n"
+       "  assign ", {kToken, "x3"}, " = 1'b0;\n"
+       "  assign ", {kToken, "x2"}, " = 1'b0;\n"
+       "  assign x1 = 1'b1;\n"
+       "endmodule"},
+
+       // generate block, TODO: multi-level
+       {"module m;\ngenerate\nendgenerate\nendmodule"},
+       {"module m;\n"
+        "  wire a1;\n"
+        "  assign a1 = 1'b1;\n"
+        "  generate\n"
+        "  endgenerate\n"
+        "  assign a1 = 1'b0;\n"
+        "endmodule"},
+       {"module m;\n"
+        "  generate\n"
+        "    wire a1;\n"
+        "    assign a1 = 1'b1;\n"
+        "  endgenerate\n"
+        "endmodule"},
+       {"module m;\n"
+        "  generate\n"
+        "    assign ", {kToken, "a1"}, " = 1'b1;\n"
+        "  endgenerate\n"
+        "endmodule"},
+       {"module m;\n"
+        "  generate\n"
+        "    wire a1;\n"
+        "    assign a1 = 1'b1;\n"
+        "  endgenerate\n"
+        "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+        "endmodule"},
+       {"module m;\n"
+        "  wire a1;\n"
+        "  assign a1 = 1'b1;\n"
+        "  generate\n"
+        "    assign ", {kToken, "a1"}, " = 1'b1;\n"
+        "  endgenerate\n"
+        "  assign a1 = 1'b0;\n"
+        "endmodule"},
+
```

Comment:
More test case ideas:

* assignments/connections inside loop and conditional generate constructs